sigstore/verify/verifier.py

```diff
@@ -133,75 +147,137 @@ def verify(
             raise VerificationError("Extended usage does not contain `code signing`")
 
-        policy.verify(bundle.signing_certificate)
+        policy.verify(cert)
 
         _logger.debug("Successfully verified signing certificate validity...")
 
-        # 4) Verify that the signature was signed by the public key in the signing certificate
+        # (4): verify the inclusion proof and signed checkpoint for the
+        #      log entry.
+        # (5): verify the inclusion promise for the log entry, if present.
+        entry = bundle.log_entry
+        try:
+            entry._verify(self._trusted_root.rekor_keyring())
+        except VerificationError as exc:
+            raise VerificationError(f"invalid log entry: {exc}")
+
+        # (6): verify that log entry was integrated circa the signing certificate's
+        #      validity period.
+        integrated_time = datetime.fromtimestamp(entry.integrated_time, tz=timezone.utc)
+        if not (
+            bundle.signing_certificate.not_valid_before_utc
+            <= integrated_time
+            <= bundle.signing_certificate.not_valid_after_utc
+        ):
+            raise VerificationError(
+                "invalid signing cert: expired at time of Rekor entry"
+            )
+
+    def verify_dsse(
+        self, bundle: Bundle, policy: VerificationPolicy
+    ) -> tuple[str, bytes]:
+        """
+        Verifies an bundle's DSSE envelope, returning the encapsulated payload
+        and its content type.
+
+        This method is only for DSSE-enveloped payloads. To verify
+        an arbitrary input against a bundle, use the `verify_artifact`
+        method.
+
+        `bundle` is the Sigstore `Bundle` to both verify and verify against.
+
+        `policy` is the `VerificationPolicy` to verify against.
+
+        Returns a tuple of `(type, payload)`, where `type` is the payload's
+        type as encoded in the DSSE envelope and `payload` is the raw `bytes`
+        of the payload. No validation of either `type` or `payload` is
+        performed; users of this API **must** assert that `type` is known
+        to them before proceeding to handle `payload` in an application-dependent
+        manner.
+        """
+
+        # (1) through (6) are performed by `_verify_common_signing_cert`.
+        self._verify_common_signing_cert(bundle, policy)
+
+        # (7): verify the bundle's signature and DSSE envelope against the
+        #      signing certificate's public key.
+        envelope = bundle._dsse_envelope
+        if envelope is None:
+            raise VerificationError(
+                "cannot perform DSSE verification on a bundle without a DSSE envelope"
+            )
+
+        dsse._verify(bundle.signing_certificate.public_key(), envelope)
+
+        # (8): verify the consistency of the log entry's body against
+        #      the other bundle materials.
+        _ = bundle.log_entry
+
+        # TODO
```

Comment:
NB: Consistency for DSSE bundles isn't implemented, pending a better understanding around how the inserted entry's `envelopeHash` is computed.